src/channel.js

```diff
@@ -0,0 +1,244 @@
+'use strict'
+
+const pushable = require('pull-pushable')
+const looper = require('looper')
+
+const { Types } = require('./consts')
+const EE = require('events')
+
+const debug = require('debug')
+
+const log = debug('pull-plex:chan')
+log.err = debug('pull-plex:chan:err')
+
+/**
+ * @fires Channel#close
+ * @fires Channel#error
+ */
+class Channel extends EE {
+  /**
+   * @constructor
+   * @param {Object} opts
+   * @param {number} opts.id
+   * @param {boolean} opts.initiator
+   * @param {string} opts.name
+   * @param {boolean} opts.open
+   * @param {Mplex} opts.plex
+   */
+  constructor (opts) {
+    super()
+
+    opts = { initiator: false, ...opts }
+
+    this._id = opts.id
+    this._name = opts.name
+    this._plex = opts.plex
+    this._open = opts.open
+    this._initiator = opts.initiator
+    this._endedRemote = false // remote stream ended
+    this._endedLocal = false // local stream ended
+    this._reset = false
+
+    this.MSG = this._initiator
+      ? Types.OUT_MESSAGE
+      : Types.IN_MESSAGE
+
+    this.END = this._initiator
+      ? Types.OUT_CLOSE
+      : Types.IN_CLOSE
+
+    this.RESET = this._initiator
+      ? Types.OUT_RESET
+      : Types.IN_RESET
+
+    this._log('new channel', this._name)
+
+    this._msgs = pushable((err) => {
+      this._log('source closed', err)
+      if (err && typeof err !== 'boolean') {
+        setImmediate(() => this.emit('error', err))
+      }
+      // this.endChan() // TODO: do not uncomment this, it will end the channel too early
+    })
+
+    this._source = this._msgs
+
+    this.sink = (read) => {
+      const next = looper(() => {
+        read(null, (end, data) => {
+          // stream already ended
+          if (this._endedLocal) { return }
+
+          this._endedLocal = end || false
+
+          // source ended, close the stream
+          if (end === true) {
+            return this.endChan()
+          }
+
+          // source errored, reset stream
+          if (end || this._reset) {
+            this.resetChan()
+            this.emit('error', end || this._reset)
+            this.reset()
+            return
+          }
+
+          // just send
+          this.sendMsg(data)
+          next()
+        })
+      })
```

Comment:
The above creates a lot of closures. I'd recommend to refactor this logic into a top level or instance function to ease V8 optimizer job.